### ISSUE TITLE
Add object position interpolation between physics frames

### ISSUE THREAD

**Issue by AndreaCatania** [godotengine/godot-proposals#671]
**Describe the problem or limitation you are having in your project:**
Usually the rendering  is much faster than the physics frames and 144hz monitors are cheap enough to be a standard device for gaming PCs nowadays.

In Godot, the physics updates the position of almost anything, usually at fixed rate of 60Hz, this mean that no matter how faster your machine is to process a frame the object rendered will change at fixed rate of 60 frame per seconds.

In other words, even with high frame rate, the new rendered frame will be the same of the previous until the next physics frame change object positions (computation waste).

The result is that high frame rate monitors are useless and the rendering is not fluid as you would expect.

However, you may want to lower the physics frame rate from 60 to 30, because your game doesn't need such precision and so to unload the CPU; doing so you will notice that the game is not fluid and you are forced to keep using 60 frames per seconds.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
This limitation can be solved by integrating the position interpolation between the frames.

We know that the physics produces a new position at a rate of 60 frames per second and between those frames the renderer produces other three identical frames. The idea is to interpolate between the old position produced by the physics engine and the new so that the intermediate frames are different each other. The problem with such approach is that we need to know the rendering speed beforehand and since this is not always stable we need a way to determine it.

The idea is to count the amount of intermediate rendering frames between the old physics update and the current one, and spread the interpolation delta between those; so even during the phases where the frame rate is not perfectly stable we are able to interpolate the position in a plausible manner.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

![Interpolation](https://user-images.githubusercontent.com/8342599/78527448-288c9100-77dd-11ea-8573-beeeac19f4f7.png)

### Without interpolation
![Interpolation_2](https://user-images.githubusercontent.com/8342599/78531900-a012ee00-77e6-11ea-8cb0-177fb375d164.png)

### With interpolation
![Interpolation_3](https://user-images.githubusercontent.com/8342599/78531904-a1dcb180-77e6-11ea-9ea6-95bb8e23541a.png)

**Comment by groud** [comment]
This looks like extrapolation then. 
Don't you fear that we could end up with weird behaviors at collision time ? Or with any wrongly extrapolated movement ? If an object enters another and go outside of it at the next physics frame we might have people complaining about some sort of stuttering.

Otherwise the idea looks good to me.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/30791. reduz is still against having it in core last time I checked, so lawnjelly's [smoothing add-on](https://github.com/lawnjelly/smoothing-addon) is our best bet for now.

**Comment by AndreaCatania** [comment]
I've updated the images to better visualize it, but let me explain better what I mean.

Godot is already rendering 1 physics frame behind this mean that what you are seeing is already the past. Instead of hardly set the new position what I'm proposing is to interpolate it between the old and the new one.

We are still 1 physics frame behind, but now we interpolate between the old and the new position.

![Interpolation_3](https://user-images.githubusercontent.com/8342599/78532222-26c7cb00-77e7-11ea-9259-6b1a4f6f632e.png)
So in the `frame 1` we already have the position 1 but instead of setting it directly to it we interpolate it.

In `frame 4` we will have the `position 2` and we will interpolate between it.

Overshooting can still happen, during frame drop, but since we are using the previous frame it's right away updated in the new iterations.

**Comment by AndreaCatania** [comment]
@Calinou yes, we should provide an API to teleport the objects or to move the object smoothly. However, this problem should be fixed because the rendering is laggy even with 120frames rendered  per second.

> Keep in mind doing this increases latency in a noticeable manner. Even if you use physics interpolation, I'd recommend not doing this unless you have absolutely no other option.

This is exactly the problem that I'm trying to solve. I can notice this high latency with my 144hz monitor and physics 60hz which is the same ratio of monitor 60Hz and physics 30Hz.

The solution for me is not increase the physics frames to 120hz so to balance it with the rendering, but rather provide a smoothing position update.

**Comment by groud** [comment]
> Godot is already rendering 1 physics frame behind this mean that what you are seeing is already the past. Instead of hardly set the new position what I'm proposing is to interpolate it between the old and the new one.

Ah, indeed in that case, this is interpolation.

Another question is how to handle things that aren't managed by the physics engine itself. I guess that for rigidbodies, the interpolation would be built-in, but for variables set by the user in `_physics_process()`, we should probably design an API to be able to smooth those properties too.

**Comment by AndreaCatania** [comment]
@groud The user script final result is something that moves in the screen, so just interpolate the result (the object position) should be enough.

**Comment by lawnjelly** [comment]
This is often referred to as fixed timestep interpolation, see:
https://github.com/godotengine/godot/issues/30068

See my article here too, it has more references:
https://www.gamedev.net/blogs/entry/2265460-fixing-your-timestep-and-evaluating-godot/

I added this PR https://github.com/godotengine/godot/pull/30226 in middle of last year to provide basic support for this, and you will notice you have the `Engine.get_physics_interpolation_fraction()` function available.

I wrote a c++ module to do it:
https://github.com/lawnjelly/godot-smooth
and a gdscript addon:
https://github.com/lawnjelly/smoothing-addon

I also researched adding it to core, and actually got it working as part of Spatial, among other things (which was quite tricky to retrofit, due to the design of Spatial, hence why I went for separate nodes in the module / addon).

However just to warn you, reduz was against adding it (in any form) to core (at the time), although he was ok with adding the interpolation_fraction PR, because the change was quite small and it allowed users to do the interpolation (without this there is no way to accurately calculate the interpolation fraction in user code).

Related .. I've also got delta smoothing working in my own branch both fixed function and also custom function from gdscript, although haven't made a PR because of lack of interest:
https://github.com/godotengine/godot/issues/30791

You've also seen the semi-fixed timestep PR which is an alternative solution and has been available since July 2019:
https://github.com/godotengine/godot/pull/30798

To clarify, I tend to use fixed timestep interpolation myself although semi-fixed can be useful in some cases, and I would think it useful to be available too.

**Comment by Jummit** [comment]
> to get it in core you'd need to get enough 'people power' to convince reduz

I think the main reason this isn't core is that no one implemented it yet or even has the knowledge how to implement it. Is someone where to propose a working solution, I don't think Reduz would be against it.

**Comment by lawnjelly** [comment]
> > to get it in core you'd need to get enough 'people power' to convince reduz
> 
> I think the main reason this isn't core is that no one implemented it yet or even has the knowledge how to implement it. Is someone where to propose a working solution, I don't think Reduz would be against it.

No, this is incorrect. It was discussed at the time, check the IRC logs.

**Comment by 2plus2makes5** [comment]
I did this myself and the result is much smoother even on a normal 60fps monitor, before that i always had an annoying stutter even on high framerates.

**Comment by aaronfranke** [comment]
Why not just increase the physics FPS? Godot allows you to do this. If the concern is with performance, wouldn't object interpolation also be an intensive operation?

Also, interpolating only the position wouldn't fully solve the problem. There is also the matter of interpolating rotations, for a rotating body.

**Comment by lawnjelly** [comment]
> Why not just increase the physics FPS? Godot allows you to do this.

There are a few reasons why changing the physics tick rate can be a bad idea, but that kind of thing is used as a strategy in some games.

1) Changing physics tick rate can change behaviour - break mechanics, jump distances etc
2) It can result in non-deterministic (ish) gameplay

Changing physics tick rate is also possible with semi-fixed timestep. I have a PR that will do this:
https://github.com/godotengine/godot/pull/30798

Also fixed timestep with interpolation is often the choice in networked games, especially server authoritative. Having different players with different tick rates could be problematic.

> If the concern is with performance, wouldn't object interpolation also be an intensive operation?

Interpolation is usually MUCH, much cheaper than physics. Orders of magnitude. It's practically free compared to rendering overheads.

> Also, interpolating only the position wouldn't fully solve the problem. If the goal is to make things smoother, there is also the matter of interpolating rotations, for a rotating body.

Yeah you have to interpolate most things. See my addon, it does basis interpolation or quaternion interpolation depending on mode.

> Also, interpolating between 2 frames requires the next frame be calculated already, so the only way to plausibly do this in real time is for the displayed objects to always be 1 physics frame behind. This adds latency, and for some use cases this is just as bad as the physics FPS being half.

Yup, this is one of the trade-offs. You can use extrapolation too, but I generally prefer interpolation. Some games use semi-fixed timestep as an alternative to get faster response.

**Comment by AndreaCatania** [comment]
I totally agree with @lawnjelly. The physics

**Comment by AndreaCatania** [comment]
Accidentally hit the Close button, sorry 🙏. 

I totally agree with @lawnjelly. The physics frame requires much much more processing power than interpolation. 

Just think about that some collision detection algorithm (the box) perform some matrix manipulation, to put the other object in the box coordinate system, before doing some other matrix manipulation to check separation axis.
this is done per each object near a box. 

Also, consider that when godot detects that an object got moved by the physics engine, it recalculates all the transformations of all child of a rigid body. 

I could continue by add the custom code in the _physics process, that is usually equally slow, etc.. 
And this is only a little part of what is needed to rendere a physics frame. So interpolation is much much less performance critical. 

Despite performance, the real problem is not solved increasing physics frames to 144hz. In case there is a good pc with a good monitor (250hz) the physics is still to slow at (144hz),and the game would not be fluid. 
In case of low end pc, the game would be too slow since the physics is running at really high speed.
In case of mobile games, or turn based games, the physics should be run at 30hz to make the game more accessible; and this should not make the objects to move jerky. 

In conclusion, the physics and the rendering, should be completely decoupled and one must not be dependent from the other for the best result always.

**Comment by AndreaCatania** [comment]
> Also, interpolating between 2 frames requires the next frame be calculated already, so the only way to plausibly do this in real time is for the displayed objects to always be 1 physics frame behind. This adds latency, and for some use cases this is just as bad as the physics FPS being half.

Just forgot to mention that Godot is already 1 physics frame behind.

**Comment by aaronfranke** [comment]
> Just forgot to mention that Godot is already 1 physics frame behind.

Interpolation needs two frames, which means the most recent frame and the previous one. If Godot is already 1 physics frame behind, then that would be the most recent frame, and then interpolation would make it up to 2 frames behind, if I'm understanding this correctly.

**Comment by AndreaCatania** [comment]
Yes, I mean that for a bug we are already another frame behind: https://github.com/godotengine/godot/issues/37702 than the normal 1 frame. So it's not so noticeable as you would expect.

**Comment by AndreaCatania** [comment]
The other day, discussing with @lawnjelly, we come with some ideas to integrate the frame interpolation in a transparent way that doesn't require a huge engine change nor the addition of the `teleport` API.

First thing first, this issue is about decoupling physics and rendering.
To achieve this we can introduce an _interpolator_ between the `PhysicsBody` node and the `PhysicsServer`. The change would remain localized and, most important, we don't need to add the `teleport` API since only the physics movement are interpolated between frames.

The _interpolator_ can submit interpolated transforms directly into the function `_direct_state_change` https://github.com/godotengine/godot/blob/master/scene/3d/physics_body_3d.cpp#L358 . The renderer will receive the interpolated position without knowing nothing about it. 

The function `get_global_transform` would return the not interpolated transform as is now.
The function `set_global_transform` would teleport the body as is now.

As you can see it has 0 API change, it's completely transparent, and it will provide us the physics interpolation.

**Comment by phantomdesvin** [comment]
> The other day, discussing with @lawnjelly, we come with some ideas to integrate the frame interpolation in a transparent way that doesn't require a huge engine change nor the addition of the `teleport` API.
> 


Godot really really needs this. I can only think that the majority of users are making games in their 60Hz monitors without knowing that they look really jittery on all 75, 144, and 240Hz monitors. I, as a begginer, had to ask the teacher of my course why his project was so jittery and he didn't seem to know so I thought that the teacher wasn't very skilled until I saw that every official demo was jittery on my PC. Only setting my physics tick rate to 144 gives me a smooth experience but that is not ideal as it is not performant and also doesn't adapt to each player. I love the engine so all my hopes are in this idea right now.

**Comment by AndreaCatania** [comment]
```cpp
// This work has been kindly sponsored by IMVU.

#ifndef FRAME_INTERPOLATOR_H
#define FRAME_INTERPOLATOR_H

#include "core/local_vector.h"
#include "core/math/math_defs.h"
#include "core/math/math_funcs.h"
#include "core/object.h"
#include "core/print_string.h"

template <class T>
class FrameInterpolator {

	LocalVector<T> data;
	T current_data;
	real_t percentage_bank = 0.0;
	real_t fixed_iter_second = 60.0;
	real_t inverse_fixed_iter_second = 1.0 / fixed_iter_second;

public:
	/// Reset the frame interpolation. Call this whenever the data store are
	/// outdated.
	void reset(uint32_t p_fixed_iteration_per_second, const T &p_current);

	/// Push the next information to interpolate. Must be called each
	/// physics_process.
	void push_data(const T &p_data);

	/// Returns the interpolated data. This must be called each `process`.
	T get_next_frame_data(const real_t p_delta);

private:
	// This removes N elements, starting from the begining.
	void range_removal(uint32_t p_remove_size);
};

template <class T>
void FrameInterpolator<T>::reset(
		uint32_t p_fixed_iteration_per_second,
		const T &p_current) {

	data.clear();
	fixed_iter_second = MAX(p_fixed_iteration_per_second, 1.0);
	inverse_fixed_iter_second = 1.0 / fixed_iter_second;
	current_data = p_current;
}

template <class T>
void FrameInterpolator<T>::push_data(const T &p_data) {
	data.push_back(p_data);
}

template <class T>
T FrameInterpolator<T>::get_next_frame_data(const real_t p_delta) {
	// The minimum buffer size is calculated so the dynamic frame delta is
	// behind of the fixed time at least of 1 unit.
	const int min_buffer_size =
			MAX(1, Math::ceil(p_delta / inverse_fixed_iter_second));

	const int offsetted_data_size = int(data.size()) - min_buffer_size;

	real_t frame_percentage = p_delta / inverse_fixed_iter_second;

	// Due to the the various frame oscillations, may happen that the buffer
	// size {`data.size()`} keep growing so the interpolated data starts become
	// old.
	// By speeding up the interpolation a bit, it's possible to avoid such case,
	// however an hard reset mechanism exist in case the buffer increses too
	// much to be easily recoverable.
	if (unlikely(offsetted_data_size > fixed_iter_second * 0.2)) {
		// Threshold for hard reset.
		// Forces the removal of all the old frames.
		percentage_bank = data.size();
	} else {
		// Speedup the 20% of fixed iterations per seconds depending how far
		// the interpolation timeline is.
		frame_percentage +=
				inverse_fixed_iter_second *
				0.2 *
				(offsetted_data_size / fixed_iter_second);
	}

	percentage_bank += frame_percentage;
	const int old_frames = percentage_bank;

	// Drop all the old frames but leave the buffer size.
	const uint32_t frames_to_drop = MIN(old_frames, MAX(offsetted_data_size, 0));
	range_removal(frames_to_drop);

	// At this point all the old frames got removed, except the buffer size.
	// By removing the integral part from the bank, the timeline is resetted
	// so the interpolation starts from the first frames of the buffer.
	percentage_bank -= old_frames;

	if (data.size() < 1) {
		return current_data;
	}

	if (frames_to_drop) {
		// There was an advances in the timeline, so just use what in the bank
		// to preven extrapolation.
		frame_percentage = percentage_bank;
	}

	current_data = current_data.interpolate_with(data[0], frame_percentage);
	return current_data;
}

template <class T>
void FrameInterpolator<T>::range_removal(uint32_t p_remove_size) {
#ifdef DEBUG_ENABLED
	// The algorithm that calls this function make sure this condition is never
	// triggered.
	CRASH_COND(p_remove_size > data.size());
#endif
	// Shift the data from the right to left, and resize buffer. This avoids
	// a lot of reallocations in case there are a lot of data to remove.
	for (uint32_t x = 0, i = p_remove_size; i < data.size(); i += 1, x += 1) {
		data[x] = data[i];
	}
	data.resize(data.size() - p_remove_size);
}

class TransformFrameInterpolator : public Object {
	GDCLASS(TransformFrameInterpolator, Object);

	FrameInterpolator<Transform> interpolator;

public:
	static void _bind_methods();

	TransformFrameInterpolator();

	void reset(uint32_t p_fixed_iteration_per_second, const Transform &p_current);
	void push_data(const Transform &p_data);
	Transform get_next_frame_data(const real_t p_delta);
};
#endif
```

The above code is the implementation of the utility that I'm using to interpolate the position of RigidBody so to unbound the physics and the rendering. The below one is the change that I did to the RigidBody in order to integrate it:

```diff
diff --git a/scene/3d/physics_body.cpp b/scene/3d/physics_body.cpp
index 828d7d08bb..a5d5862bdc 100644
--- a/scene/3d/physics_body.cpp
+++ b/scene/3d/physics_body.cpp
@@ -470,8 +470,7 @@ void RigidBody::_direct_state_changed(Object *p_state) {
        state = (PhysicsDirectBodyState *)p_state; //trust it
 #endif
 
-       set_ignore_transform_notification(true);
-       set_global_transform(state->get_transform());
+       frame_interpolator.push_data(state->get_transform());
        linear_velocity = state->get_linear_velocity();
        angular_velocity = state->get_angular_velocity();
        if (sleeping != state->is_sleeping()) {
@@ -480,7 +479,6 @@ void RigidBody::_direct_state_changed(Object *p_state) {
        }
        if (get_script_instance())
                get_script_instance()->call("_integrate_forces", state);
-       set_ignore_transform_notification(false);
 
        if (contact_monitor) {
 
@@ -572,20 +570,38 @@ void RigidBody::_direct_state_changed(Object *p_state) {
 
 void RigidBody::_notification(int p_what) {
 
-#ifdef TOOLS_ENABLED
-       if (p_what == NOTIFICATION_ENTER_TREE) {
-               if (Engine::get_singleton()->is_editor_hint()) {
-                       set_notify_local_transform(true); //used for warnings and only in editor
-               }
-       }
+       switch (p_what) {
+               case NOTIFICATION_INTERNAL_PROCESS: {
 
-       if (p_what == NOTIFICATION_LOCAL_TRANSFORM_CHANGED) {
-               if (Engine::get_singleton()->is_editor_hint()) {
-                       update_configuration_warning();
-               }
-       }
+                       const real_t delta = get_process_delta_time();
+
+                       set_ignore_transform_notification(true);
+                       set_global_transform(frame_interpolator.get_next_frame_data(delta));
+                       set_ignore_transform_notification(false);
 
+               } break;
+               case NOTIFICATION_TRANSFORM_CHANGED:
+               case NOTIFICATION_READY:
+                       if (Engine::get_singleton()->is_editor_hint() == false) {
+                               set_process_internal(true);
+                               frame_interpolator.reset(
+                                               Engine::get_singleton()->get_iterations_per_second(),
+                                               get_global_transform());
+                       }
+                       break;
+#ifdef TOOLS_ENABLED
+               case NOTIFICATION_ENTER_TREE:
+                       if (Engine::get_singleton()->is_editor_hint()) {
+                               set_notify_local_transform(true); //used for warnings and only in editor
+                       }
+                       break;
+               case NOTIFICATION_LOCAL_TRANSFORM_CHANGED:
+                       if (Engine::get_singleton()->is_editor_hint()) {
+                               update_configuration_warning();
+                       }
+                       break;
 #endif
+       }
 }
```

As you can see, the required modification is completely contained into the RigidBody and it's possible to have the following result **Physics 10Hz - interpolation**:
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8342599/94676364-8547af00-031b-11eb-9d53-4bfcc8e7b01f.gif)

That can be compared to, **Physics 10Hz - no interpolation**:
![ezgif com-video-to-gif(1)](https://user-images.githubusercontent.com/8342599/94676628-ee2f2700-031b-11eb-8df5-7c665007e4f4.gif)

Side by side comparison: https://youtu.be/En-zhIU8zbI

Note: Teleport is fully handled transparently.
Note: I've noticed that not always the interpolation is wanted. For example, a Kinematic Body used in a networked game. So in case we would implement this, we would need a way to deactivate the interpolation for a specific RigidBody / Kinematic Body.

**Comment by swift502** [comment]
Can this be added to the 4.0 milestone so it's not forgotten about?

**Comment by AndreaCatania** [comment]
@swift502 Yes you have linear and angular interpolation, it happens thanks to `interpolate_with`. I've just added the CPP side too, not too much code but it was missing. Let me know if it works, now.

**Comment by swift502** [comment]
If there was a C# build with this feature available, I'd be happy to test it and confirm proper functionality in my main project.

**Comment by swift502** [comment]
Sorry to bother @pouleyKetchoupp but are there any plans to look for a solution to physics interpolation with the physics refactors you're working on for 4.0? Could we possibly stick a 4.0 or 4.x milestone on this?

Thanks for any info on this.

**Comment by pouleyKetchoupp** [comment]
@swift502 Physics interpolation is on the roadmap, although it needs more discussion within the physics team so it might be only for 4.1.

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/2753, which is more detailed and closer to the final implementation present in 3.5beta so far.
